Initialize hosting trace with OTEL tags for sampling

rkargMsft opened the pull request:
# Feature for adding tags at Activity creation time

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes

Adding `server.address` and `server.port` tags at Activity creation. This is to support using those tags when making sampling decisions as only tags provided in the `CreateActivity`/`StartActivity` are available at that time.

## Description

OpenTelemetry spec also says that `http.request.method` SHOULD be added but not including that in this PR due to the extra logic needed to compute that value.

Fixes #50488

rkarg-blizz committed:
Feature for setting activity tags on creation

rkarg-blizz committed:
adding unshipped public API entries

rkarg-blizz committed:
moving to TagList

rkarg-blizz committed:
Less specific property type

also fixed up unshipped API documentation

JamesNK commented:
What's the scenario where you want to set activity tags on creation? This seems like kind of niche feature and I'd prefer to see whether the problem can be solved with existing APIs.

For example, there is a feature to get the current request activity, `IHttpActivityFeature`, and there are many events during the corse of a HTTP request. Could you use one of those existing events, get the activity using `IHttpActivityFeature` and then set the tags?

JamesNK commented:
> This is to support using those tags when making sampling decisions as only tags provided in the CreateActivity/StartActivity are available at that time.

Won't the activity name always be present at this time? Why not use the activity name to make the sampling decision? The name will always be the same for the HTTP activity, but the specified tags from this feature will always be the same as well.

rkargMsft commented:
I'm specifically using YARP as a reverse proxy for a large number of services. There are calls for many different host names. I want to make tracing sampling decision for those calls and would like to have separate buckets for each hostname. For example, in a given time window, I want a minimum number of traces to always be sampled and then start doing probabilistic sampling after that.

As it is now, I can only get the base operation name (just `HTTP <METHOD>` at the time of sampling) and no tags. This means that if I have one very high throughput service and one very low throughput service then they're all in the same bucket for sampling (there's not host name or other data in tags yet to differentiate requests) and probabilistically, the low throughput service basically never gets sampled.

The only way to get tags to be available at the time of making the sampling decision is for them to be passed in to the `CreateActivity` or `StartActivity` call and the existing code hard codes that to `tags:null`.

The flow is:
- Create/Start Activity
  - Sampling logic determines if an Activity is returned or null based on only the parameters passed in on Create/Start
- If non-null (because sampling decided to keep the span) then additional tags or modifications can be made

I don't know of any other way to get information available to the sampler than to pass the tags in on Create/StartActivity.

This is extra work for _every_ request, even those that end up not sampled, which is why it does nothing by default and would require a user to specifically opt in with their own implementation for specifically the data points they need in their particular scenario. For example, the vast majority of our services would not use this functionality; only our YARP usage would define an implementation to just add the host/`server.address` tag.

rkargMsft commented:
/azp run

rkargMsft committed:
Update HostingApplicationDiagnostics.cs

rkargMsft commented:
@JamesNK This continues to be a problem for us, preventing us from getting useful trace sampling on our YARP load balancer. I don't see any workaround we can do without this change or something else that updates how the Activity is created internally.

Is there anything else that can help with this this considered for the .NET 10 timeframe?

rkargMsft commented:
That sounds like an accurate summary.

The existing tags added after Activity creation are made with `AddTag`
https://github.com/dotnet/aspnetcore/blob/45635b6a9dbd4abd8decb6ac00b7b2058f8c9374/src/Shared/Diagnostics/ActivityCreator.cs#L79
so if there's a Feature registered that sets a tag (like `server.address`) then that will **not** get overwritten after sampling. It will remain with whatever value the Feature populated it with.

That seems like reasonable behavior as long as it's documented.

rkargMsft commented:
An other option to consider could be hard coding an opinionated set of tags at Activity creation (like `server.address` and `server.port`) and always add those at Activity creation time.

The downsides of that are 
* this cost gets paid by everyone, regardless of whether they use tags for sampling or not. This gets paid for every Activity, even the ones that eventually get sampled out. It's unclear what a good way to opt-out would be that is better than this proposed PR would offer (with this PR being strictly opt-in)
* this set of tags may not satisfy all use cases so an extension point may still be needed anyway

JamesNK commented:
I don't want to add a new feature. I'd prefer that we followed what the standard proposes and makes the [here](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md?rgh-link-date=2025-05-26T15%3A14%3A28Z#:~:text=bots%20or%20tests.-,The%20following%20attributes%20can%20be%20important%20for%20making%20sampling%20decisions%20and%20SHOULD%20be%20provided%20at%20span%20creation%20time%20(if%20provided%20at%20all)%3A,-http.request.method) about including some tags at creation time.

The problem is there is a lot of the standard we don't follow around traces, and I doubt we'll have time in .NET 10 to switch to following it.

rkargMsft committed:
unconditionally adding host/port tags to Activity creation

This removes the need to expose a new Feature or make any other public changes.

rkargMsft commented:
If the cost of setting those tags (which should be minimal) is acceptable then those could get added.
PR updated to reflect the below proposal.

Would an incremental step be adding just:
* `server.address` 
* `server.port` 
to the tags at creation time? Those are both properties already available from the HttpContext that's passed in as an argument to HostingApplicationDiagnostics.StartActivity.

Basically replace the Feature implementation in this PR with this:
```csharp
        var tagsForCreation = new TagList();
        if (_activitySource.HasListeners() && _context.Request.Host.HasValue)
        {
            var (host, port) = _context.Request.Host.HostAndPort;
            tags.Add("server.address", host);
            if (port is not null)
            {
                tags.Add("server.port", port);
            }
            else if (string.Equals("https", httpContext.Request.Scheme, StringComparison.OrdinalIgnoreCase))
            {
                tags.Add("server.port", 443);
            }
            else if (string.Equals("http", httpContext.Request.Scheme, StringComparison.OrdinalIgnoreCase))
            {
                tags.Add("server.port", 80);
            }
        }
```

Setting the `http.request.method` would require calling HostingMetrics.ResolveHttpMethod which is currently private.

`url.full` is only defined in the HTTP Client span tags so that wouldn't apply here.

If just host/port get added then it shouldn't be a breaking change later when the full OTel compliance gets implemented (since it would just be adding the method tag with the resolved value).

There's also the opportunity to optimize the `.Host` and `.Port` access on HostString as there isn't a way to get both of those items in one call even though the internal implementation parses both of those string when you ask for either. There's an existing `GetHostAndPort` usage [in HostMatcherPolicy](https://github.com/dotnet/aspnetcore/blob/0d5ddb0e9083252f90b247dcdf64c659242a254a/src/Http/Routing/src/Matching/HostMatcherPolicy.cs#L333-L352) that could benefit from this as well as the current use case. A `(string host, int? port) HostString.HostAndPort` property would remove redundant searching for splitting the host/port.

rkargMsft committed:
correcting variable naming

rkargMsft committed:
reverting HostAndPort property

rkargMsft committed:
naming

BrennanConroy commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -389,6 +389,26 @@ private void RecordRequestStartMetrics(HttpContext httpContext)
+        {
+            var host = httpContext.Request.Host.Host;
+            var port = httpContext.Request.Host.Port;
+            tagsForCreation.Add("server.address", host);
```
This doesn't seem like the 100% right way to get the value for "server.address". Per [OTEL spec](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/registry/attributes/server.md):
> Server domain name if available without reverse DNS lookup; otherwise, IP address or Unix domain socket name. [1]

See https://github.com/dotnet/aspnetcore/blob/062e663e3f634ef1b99e347a7e7cf7de363c9f53/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs#L322 for how we do it for Kestrel metrics.

You might want https://github.com/dotnet/aspnetcore/pull/62162 which I was playing around with.

rkargMsft replied:
The Otel Contrib [AspNetCore instrumentation package](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/63071941b7da49f3efb7dd2c3b592dbc3d7268a6/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L184-L194) sets it from the context.Request.Host.Host value.

HTTP spans have specific convention for server.address: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#setting-serveraddress-and-serverport-attributes

rkargMsft replied:
/shrug

BrennanConroy commented:
This seems like a good step forward. James is more aware of the OTel space so I'll leave the final call to @JamesNK.

JamesNK committed:
Initialize hosting trace with OTEL tags for sampling

JamesNK commented:
@rkargMsft I took a look at your PR, and how it interacts with OTEL libraries. I think populating these OTEL attributes on activity creation is a good approach.

I don't think it will cause interop problems with OTEL libraries that add tags to the activity, but some more testing during RC period is required.

I've made some improvements to your PR. Adding more recommended created attributes, small perf optimizations, adding tests, adding an opt out switch incase this creates problems for anyone.

Are you ok with these changes?

martincostello commented on src/Hosting/Hosting/src/Internal/HostingTelemetryHelpers.cs:
```diff
@@ -0,0 +1,132 @@
+
+    private static readonly object[] BoxedStatusCodes = new object[512];
+
+    private static readonly FrozenDictionary<string, string> KnownHttpMethods = FrozenDictionary.ToFrozenDictionary([
+        KeyValuePair.Create(HttpMethods.Connect, HttpMethods.Connect),
+        KeyValuePair.Create(HttpMethods.Delete, HttpMethods.Delete),
+        KeyValuePair.Create(HttpMethods.Get, HttpMethods.Get),
+        KeyValuePair.Create(HttpMethods.Head, HttpMethods.Head),
+        KeyValuePair.Create(HttpMethods.Options, HttpMethods.Options),
+        KeyValuePair.Create(HttpMethods.Patch, HttpMethods.Patch),
+        KeyValuePair.Create(HttpMethods.Post, HttpMethods.Post),
+        KeyValuePair.Create(HttpMethods.Put, HttpMethods.Put),
+        KeyValuePair.Create(HttpMethods.Trace, HttpMethods.Trace)
+    ], StringComparer.OrdinalIgnoreCase);
```
Add `QUERY` given #63260 went in yesterday?

JamesNK replied:
This list is based on the known values list at https://opentelemetry.io/docs/specs/semconv/registry/attributes/http

<img width="1317" height="849" alt="image" src="https://github.com/user-attachments/assets/d1ce3ec3-fa56-41a5-bb59-21ba96381708" />

You're welcome to get OTEL folks to add QUERY and then we'll add it here.

rkargMsft commented:
This looks good. We can also run the RCs and provide feedback.

JamesNK commented:
In case this change causes problems or we decide that it's too risky, @rkargMsft what do you think of defaulting it to off and using the AppContext switch to turn it on?

i.e. in .NET 10 by default there are no creation tags and to get them you would add `AppContext.SetSwitch("Microsoft.AspNetCore.Hosting.SuppressActivityOpenTelemetryData", false)` to your ASP.NET Core app startup code. Would that be acceptable in your app(s)?

JamesNK committed:
Default to off

JamesNK commented:
/backport to release/10.0-rc1
